### DISCUSSION TITLE
APSIM build fails when creating a singularity container

### DISCUSSION THREAD

**Discussion by BrianCollinss** [APSIMInitiative/ApsimX#8732] [Q&A] [answered]
We get this error when we try to build a singularity using the APSIM code that resides in a fork/branch. Any idea what the cause of it may be?

![BCApsimXError (002)](https://github.com/APSIMInitiative/ApsimX/assets/39573622/c3e70975-b890-42b1-a295-4007a787f993)

**Comment by par456** [comment]
That must be an older fork, I just had a look at the UnitTests.csproj file from the main repo and it's set to version 17.5.0

My suggestion would be to do a search through these files (if they exist on that fork):
Tests/UnitTests.csproj
Tests/UnitTests/obj/project.assets.json
Tests/UnitTests/obj/UnitTests.csproj.nuget.dgspec.json

And change all references to Microsoft.NET.Test.Sdk to the same version (use 16.6.1 since that seems to be the highest version on that fork).

**Comment by BrianCollinss** [comment]
@par456 Thanks Andrew. Could you please help me better understand how forks work? What I do usually is push my changes to a branch of this work. Then open a pull request, it is reviewed and perhaps merged. Now, I want to build APSIM from my fork (with or without my recent changes having been merged or not). What should I do? Do I need to push those changes to the master branch of my fork?

**Reply by par456** [reply, marked as answer]
Your fork will have it's own main branch that most likely isn't up to date if you've been getting your branches merged back to the main apsim repo with pull requests (as you might not have been merging those changes back into your main branch).

So you'll need to do a pull and merge to get your main branch up to date with the main repo again. If it can merge successfully, that should allow you to build apsim with all the latest changes from your own fork. You can then also merge your branches into your own fork without it going into the main apsim repo.

The main reason to do this is so that you can merge branches together into your main to access changes from multiple branches without having them merged back into the main repo. However, if your fork gets too far from the apsim main branch, it may become impossible to merge all the changes back in, so do take note of that. You nearly always want to work on a branch and merge your changes back into your main to test, and not work directly on main.
